### PR TITLE
Add rough support for chrono's Naive{DateTime,Date,Time} types

### PR DESCRIPTION
**Pull request by Eijebong**
This is #344 with tests and fixed date parsing

I took the liberty to finish that since nothing happened in the PR for more than a year.

I had to parse dates without chrono since %Y doesn't support years > 9999

### CODE REVIEW

File: diesel/src/sqlite/types/date_and_time/chrono.rs [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+    fn to_sql<W: Write>(&self, out: &mut W) -> Result<IsNull, Box<Error+Send+Sync>> {
+        let s = dump_sqlite_date(*self);
+        ToSql::<Text, Sqlite>::to_sql(&s, out)
+    }
+}
+
+impl FromSql<Time, Sqlite> for NaiveTime {
+    fn from_sql(value: Option<&SqliteValue>) -> Result<Self, Box<Error+Send+Sync>> {
+        let text = not_none!(value).read_text();
+        let time = try!(NaiveTime::parse_from_str(text, SQLITE_TIME_FMT));
```
Comment (killercup) [inline diesel/src/sqlite/types/date_and_time/chrono.rs]: use `?` for consistency :)

File: diesel/src/sqlite/types/date_and_time/chrono.rs [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+use sqlite::Sqlite;
+use sqlite::connection::SqliteValue;
+use types::{Date, FromSql, IsNull, Time, Timestamp, ToSql, Text};
+
+const SQLITE_TIME_FMT: &'static str = "%0H:%0M:%0S%.6f";
+
+fn parse_sqlite_date(date: &str) -> Result<NaiveDate, Box<Error+Send+Sync>> {
+    let negative_year = date.starts_with('-');
+
+    let ymd: Vec<u32> = date.split("-").filter_map(|s| s.parse().ok()).collect();
```
Comment (Eijebong) [inline diesel/src/sqlite/types/date_and_time/chrono.rs]: Oh, I didn't know that one ^^

**Comment by Eijebong** [comment]
I just took #344, rebased, fixed what needed to be fixed and added tests.

What kind of things does it need ?

**Comment by killercup** [comment]
Off the top of my head, I'd say

- [ ] integration tests (maybe just enabling/copying pg tests) that use a struct with a Date/Time/DateTime field (i think you need `expression_impls!` and `queryable_impls!` at least)
- [ ] roundtrip tests
- [ ] mention the sqlite impl (and maybe also caveats of it) in the docs, e.g. in `diesel::types::Date`
- [ ] changelog entry

(feel free to copy this list)

File: diesel/src/sqlite/types/date_and_time/chrono.rs [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+use types::{Date, FromSql, IsNull, Time, Timestamp, ToSql, Text};
+
+const SQLITE_TIME_FMT: &'static str = "%0H:%0M:%0S%.6f";
+
+fn parse_sqlite_date(date: &str) -> Result<NaiveDate, Box<Error+Send+Sync>> {
+    let negative_year = date.starts_with('-');
+
+    let ymd: Vec<u32> = date.split('-').flat_map(str::parse).collect();
+    if ymd.len() != 3 {
+        return Err("Cannot parse date".into())
```
Comment (killercup) [inline diesel/src/sqlite/types/date_and_time/chrono.rs]: Maybe add a more precise error message here? This is the same as below, so people getting this error and searching for the string will not know which one it was.

File: diesel/src/sqlite/types/date_and_time/chrono.rs [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+    }
+
+    if negative_year {
+        return Ok(NaiveDate::from_ymd(-(ymd[0] as i32), ymd[1], ymd[2]))
+    }
+
+    match NaiveDate::from_ymd_opt(ymd[0] as i32, ymd[1], ymd[2]) {
+        Some(d) => Ok(d),
+        None => Err("Cannot parse date".into()),
```
Comment (killercup) [inline diesel/src/sqlite/types/date_and_time/chrono.rs]: Maybe add a more precise error message here?²

File: diesel/src/sqlite/types/date_and_time/chrono.rs [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+    }
+}
+
+impl FromSql<Timestamp, Sqlite> for NaiveDateTime {
+    fn from_sql(value: Option<&SqliteValue>) -> Result<Self, Box<Error+Send+Sync>> {
+        let text = not_none!(value).read_text();
+        let dt: Vec<&str> = text.split(' ').collect();
+
+        if dt.len() != 2 {
+            return Err("Can't parse Datetime".into())
```
Comment (killercup) [inline diesel/src/sqlite/types/date_and_time/chrono.rs]: Maybe add a more precise error message here?³

**Comment by Eijebong** [comment]
Fixed and rebased.

**Comment by Eijebong** [comment]
Merged. Thanks @kardeiz for the original implementation :)

**Pull request merged**